internal/proxy/proxy.go

```diff
@@ -298,9 +297,54 @@ func NewClient(ctx context.Context, d cloudsql.Dialer, l cloudsql.Logger, conf *
 	return c, nil
 }
 
+// CheckConnections dials each registered instance and reports any errors that
+// may have occurred.
+func (c *Client) CheckConnections(ctx context.Context) error {
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, len(c.mnts))
+	)
+	for _, m := range c.mnts {
+		wg.Add(1)
+		go func(inst string) {
+			defer wg.Done()
+			conn, err := c.dialer.Dial(ctx, inst)
```

Comment:
If we are doing dial and immediately close every 5 seconds to all the databases from all the pods, is this going to exceed any api request limits?

Reply:
Not necessarily. The dialer re-uses connection info under the hood and only refreshes that info every ~hour. For large deployments, though, just using the proxy can hit up against API quota.

The bigger question is if we should be dialing instances at all in the readiness check. I've read that readiness shouldn't be checking connections to downstream services. What would a good readiness check be for the proxy?

Reply:
Well, it depends on what ready means in the case of the proxy? 

When the proxy is a standalone deployment to me the choice is clear: ready means "database connection is ok." 

With proxy as a sidecar container to an application, the application won't receive traffic until both the app container and the sidecar both indicate "ready." Competent kubernetes application developers will have their application also expose a readiness check that will test the database connection before allowing traffic to the application. 

We have two choices:

- **Option A:** Proxy only says it is "ready" when the database can connect. If app developer added a database check to the application's readiness check, then the proxy readiness check is redundant. If app developer does not not add a database connection, the developer is saved from routing traffic to an application that can't connect to its database.
- **Option B:** Proxy says it is "ready" when its configuration is fully loaded and ports are open and listening. If the app developer added a readiness check, then the app won't receive traffic until the database can connect. If not, the app will become "ready" and start receiving traffic regardless of the ability to connect to the db.

Unfortunately, not every developer is a competent kubernetes application developer. Most of our customers rely on us to know "the right thing to do." So I think Option A is a safer option for most customers.

Reply:
Startup check covers "proxy has loaded the config and mounted it's listeners". 

Liveliness checks "the proxy is running"

Readiness checks are when the pod ready to accept traffic. I don't see a better way to do that than to test dial the instances.